android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt

```diff
@@ -0,0 +1,87 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.state.engine.middleware
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import mozilla.components.browser.state.action.BrowserAction
+import mozilla.components.browser.state.action.TranslationsAction
+import mozilla.components.browser.state.action.TranslationsAction.TranslateExpectedAction
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.concept.engine.Engine
+import mozilla.components.concept.engine.translate.TranslationError
+import mozilla.components.concept.engine.translate.TranslationOperation
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import mozilla.components.support.base.log.logger.Logger
+
+/**
+ * This middleware is for use with managing any states or resources required for translating a
+ * webpage.
+ */
+class TranslationsMiddleware(
+    private val engine: Engine,
+    private val scope: CoroutineScope,
+) : Middleware<BrowserState, BrowserAction> {
+    private val logger = Logger("TranslationsMiddleware")
+
+    override fun invoke(
+        context: MiddlewareContext<BrowserState, BrowserAction>,
+        next: (BrowserAction) -> Unit,
+        action: BrowserAction,
+    ) {
+        // Pre process actions
+        when (action) {
+            is TranslateExpectedAction -> {
+                requestSupportedLanguages(context, action.tabId)
+            }
+            else -> {
+                // no-op
+            }
+        }
+
+        // Continue to post process actions
+        next(action)
+    }
+
+    /**
+     * Retrieves the list of supported languages using [scope] and dispatches the result to the
+     * store via [TranslationsAction.TranslateSetLanguagesAction] or else dispatches the failure
+     * [TranslationsAction.TranslateErrorAction].
+     *
+     * @param context Context to use to dispatch to the store.
+     * @param tabId Tab ID associated with the request.
+     */
+    private fun requestSupportedLanguages(
+        context: MiddlewareContext<BrowserState, BrowserAction>,
+        tabId: String,
+    ) {
+        scope.launch {
+            engine.getSupportedTranslationLanguages(
```

Comment:
Not sure if this is a long running tasks, just bring it up that it will run by default on the main thread :) 

https://github.com/mozilla-mobile/firefox-android/blob/efd8f194c333e87d0ccc639e0b336222b1cb42a7/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/EngineMiddleware.kt#L38

Reply:
I wasn't really sure what the best practice was here, LMKWYT! 

Here is some more context:

 I don't expect it to be a longer task, but it is passing through a lot of layers. It will traverse AC <-> GV Java <-> GV JavaScript <-> Toolkit logic <-> Remote Settings. AFAIK, in Remote Settings it is querying a file with this list because the data is available offline. So, no server call with a long wait at the end, but a lot of message passing.

Reply:
Perfect, thanks for the all the clarifications, I think we could proceed as it's and adapt in the future if it's needed!

Reply:
>It isn't always true that the file is available offline, I think. Or the file may need to be re-fetched from the server before it can respond?

I double checked and what you said is correct, glad you brought it up! I did a quick test in GeckoView to check my assumptions. So, with a cold install with airplane mode turned on, there are no languages. I looked through the code and I think the very first time it does do a server query. Any second attempt, e.g., fetch once, close up completely, reopen in airplane mode, will have cached  languages. I had only tested the second case in the past.

I updated to `private val scope: CoroutineScope = CoroutineScope(Dispatchers.IO)` as a precaution.